src/common/string/filter/filter.ts

```diff
@@ -0,0 +1,448 @@
+ * 0. the score
+ * 1. the matches encoded as bitmask (2^53)
+ * 2. the offset at which matching started
+ */
+export type FuzzyScore = [number, number, number];
+
+export function fuzzyScore(
+  pattern: string,
+  patternLow: string,
+  patternStart: number,
+  word: string,
+  wordLow: string,
+  wordStart: number,
+  firstMatchCanBeWeak: boolean
+): FuzzyScore | undefined {
+  const patternLen = pattern.length > _maxLen ? _maxLen : pattern.length;
+  const wordLen = word.length > _maxLen ? _maxLen : word.length;
+
+  if (
+    patternStart >= patternLen ||
+    wordStart >= wordLen ||
+    patternLen - patternStart > wordLen - wordStart
+  ) {
+    return undefined;
+  }
+
+  // Run a simple check if the characters of pattern occur
+  // (in order) at all in word. If that isn't the case we
+  // stop because no match will be possible
+  if (
+    !_isPatternInWord(
+      patternLow,
+      patternStart,
+      patternLen,
+      wordLow,
+      wordStart,
+      wordLen
+    )
+  ) {
+    return undefined;
+  }
+
+  let row = 1;
+  let column = 1;
+  let patternPos = patternStart;
+  let wordPos = wordStart;
+
+  let hasStrongFirstMatch = false;
+
+  // There will be a match, fill in tables
+  for (
+    row = 1, patternPos = patternStart;
+    patternPos < patternLen;
+    row++, patternPos++
+  ) {
+    for (
+      column = 1, wordPos = wordStart;
+      wordPos < wordLen;
+      column++, wordPos++
+    ) {
+      const score = _doScore(
+        pattern,
+        patternLow,
+        patternPos,
+        patternStart,
+        word,
+        wordLow,
+        wordPos
+      );
+
+      if (patternPos === patternStart && score > 1) {
+        hasStrongFirstMatch = true;
+      }
+
+      _scores[row][column] = score;
+
+      const diag = _table[row - 1][column - 1] + (score > 1 ? 1 : score);
+      const top = _table[row - 1][column] + -1;
+      const left = _table[row][column - 1] + -1;
+
+      if (left >= top) {
+        // left or diag
+        if (left > diag) {
+          _table[row][column] = left;
+          _arrows[row][column] = Arrow.Left;
+        } else if (left === diag) {
+          _table[row][column] = left;
+          _arrows[row][column] = Arrow.Left || Arrow.Diag;
+        } else {
+          _table[row][column] = diag;
+          _arrows[row][column] = Arrow.Diag;
+        }
+      } else if (top > diag) {
+        _table[row][column] = top;
+        _arrows[row][column] = Arrow.Top;
+      } else if (top === diag) {
+        _table[row][column] = top;
+        _arrows[row][column] = Arrow.Top || Arrow.Diag;
+      } else {
+        _table[row][column] = diag;
+        _arrows[row][column] = Arrow.Diag;
+      }
+    }
+  }
+
+  if (_debug) {
+    printTables(pattern, patternStart, word, wordStart);
+  }
+
+  if (!hasStrongFirstMatch && !firstMatchCanBeWeak) {
+    return undefined;
+  }
+
+  _matchesCount = 0;
+  _topScore = -100;
+  _wordStart = wordStart;
+  _firstMatchCanBeWeak = firstMatchCanBeWeak;
```

Comment:
I hate their use of globals, makes it much less readable. Would also blow up when we would use it async?

Reply:
Hmm. Yeah, fair point. I'll see if I can work the vars into the recursion

Reply:
Done